File: event-service/src/main/java/com/project/event/config/RabbitConfig.java
package com.project.event.config;

import org.springframework.amqp.core.*;
import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitAdmin;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class RabbitConfig {
    public static final String USER_SERVICE_QUEUE = "event-service";
    @Value("${rabbit.event-service.exchange}")
    private String EXCHANGE;
    @Value("${rabbit.evnt-service.routingKey}")
    private String ROUTING_KEY;

    @Value("${spring.rabbitmq.host}")
    private String host;

    @Bean
    public ConnectionFactory connectionFactory() {
        return new CachingConnectionFactory(host);
    }

    @Bean
    public AmqpAdmin admin() {
        return new RabbitAdmin(connectionFactory());
    }

    @Bean
    public RabbitTemplate rabbitTemplate() {
        var template = new RabbitTemplate(connectionFactory());
        template.setMessageConverter(new Jackson2JsonMessageConverter());
        return template;
    }

    @Bean
    public Queue userServiceQueue() {
        return new Queue(USER_SERVICE_QUEUE, true, false, false);
    }

    @Bean
    public TopicExchange exchange() {
        return new TopicExchange(EXCHANGE);
    }

    @Bean
    public Binding binding(TopicExchange exchange) {
        return BindingBuilder.bind(userServiceQueue()).to(exchange).with(ROUTING_KEY);
    }
}

File: notification-service/src/main/kotlin/com/project/notificationService/config/RabbitConfig.kt
package com.project.notificationService.config

import org.springframework.amqp.core.*
import org.springframework.amqp.rabbit.connection.CachingConnectionFactory
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.rabbit.core.RabbitAdmin
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class RabbitConfig {
    @Value("\${spring.rabbitmq.host}")
    private lateinit var host: String
    @Value("\${rabbit.notification-service.exchange}")
    private lateinit var EXCHANGE: String

    @Value("\${rabbit.notification-service.routingKey}")
    private lateinit var ROUTING_KEY: String

    @Bean
    fun connectionFactory(): ConnectionFactory {
        return CachingConnectionFactory(host)
    }

    @Bean
    fun admin(): AmqpAdmin {
        return RabbitAdmin(connectionFactory())
    }

    @Bean
    fun rabbitTemplate(): RabbitTemplate {
        val template = RabbitTemplate(
            connectionFactory()
        )
        template.messageConverter = Jackson2JsonMessageConverter()
        return template
    }

    @Bean
    fun userServiceQueue(): Queue {
        return Queue(NOTIFICATION_SERVICE_QUEUE, true, false, false)
    }

    @Bean
    fun exchange(): TopicExchange {
        return TopicExchange(EXCHANGE)
    }

    @Bean
    fun binding(exchange: TopicExchange): Binding {
        return BindingBuilder.bind(userServiceQueue()).to(exchange).with(ROUTING_KEY)
    }

    companion object {
        const val NOTIFICATION_SERVICE_QUEUE = "send-notification"
    }
}